lib/cli/watch-run.js

```diff
@@ -137,44 +139,226 @@ exports.watchRun = (mocha, {watchFiles, watchIgnore}, fileCollectParams) => {
   });
 };
 
-class GlobFilesTracker {
-  constructor(watchFiles, watchIgnore) {
-    this.watchFilesSet = new Set();
-    this.watchFiles = watchFiles;
-    this.watchIgnore = watchIgnore;
-  }
-
-  regenerate() {
-    const watchIgnoreSet = new Set();
-    for (const pattern of this.watchIgnore) {
-      glob.sync(pattern, { dot: true }).forEach(filePath => watchIgnoreSet.add(filePath));
-    }
+/**
+ * Extracts out paths without the glob part, the directory paths,
+ * and the paths for matching from the provided glob paths.
+ * @param {string[]} globPaths The list of glob paths to create a filter for.
+ * @param {string} basePath The path where mocha is run (e.g., current working directory).
+ * @returns {PathFilter} Object to filter paths.
+ * @ignore
+ * @private
+ */
+function createPathFilter(globPaths, basePath) {
+  debug('creating path filter from glob paths: %s', globPaths);
+
+  /**
+   * The resulting object to filter paths.
+   * @type {PathFilter}
+   */
+  const res = {
+    dir: {paths: new Set(), globs: new Set()},
+    match: {paths: new Set(), globs: new Set()}
+  };
 
-    const globOpts = {
+  // for checking if a path ends with `/**/*`
+  const globEnd = path.join(path.sep, '**', '*');
+  
+    /**
+     * The current glob pattern to check.
+     * @type {typeof Pattern[]} The `Pattern` class is not exported by the `glob` package.
+     * Ref [link](../../node_modules/glob/dist/commonjs/pattern.d.ts)
+     */
+  const patterns = globPaths.reduce((acc, globPath) => {
+    const globInstance = new glob.Glob(globPath, {
       dot: true,
-      ignore: {
-        childrenIgnored: pathToCheck => watchIgnoreSet.has(pathToCheck.relative())
+      magicalBraces: true,
+      windowsPathsNoEscape: true
+    });
+    return acc.concat(globInstance.patterns);
+  }, []);
+
+  // each pattern will have its own path because of the `magicalBraces` option
+  for (const pattern of patterns) {
+    debug('processing glob pattern: %s', pattern.globString());
+
+    /**
+     * Path segments before the glob pattern.
+     * @type {string[]}
+     */
+    const segments = [];
+
+    /**
+     * The current glob pattern to check.
+     * @type {typeof Pattern | null} The `Pattern` class is not exported by the `glob` package.
+     */
+    let currentPattern = pattern;
+    let isGlob = false;
+
+    do {
+      // save string patterns until a non-string (glob or regexp) is matched
+      const entry = currentPattern.pattern();
+      const serializedEntry = JSON.stringify(entry);
+      debug('processing glob subpattern: %s', serializedEntry);
```

Comment:
You can use `%j` in place of `%s`/`JSON.stringify` here

Reply:
Nice, good to know moving forward, others are welcome to make this change but since it only affects debug builds I'm not gonna prioritize it :D